R/fct_SQLite.R

```diff
@@ -488,38 +474,38 @@ db_get_query <- function(
 #'   temp_path <- withr::local_tempfile(fileext = ".sqlite")
 #'   con <- get_db_connection(temp_path)
 #'   review_data <- data.frame(
-#'   subject_id = "Test_name",
-#'    event_name = "Visit 1",
-#'    item_group = "Test_group",
-#'    form_repeat = 1,
-#'    item_name = "Test_item",
-#'    edit_date_time = "2023-11-05 01:26:00",
-#'    timestamp = "2024-02-05 01:01:01"
+#'    subject_id = c("Test_name", "Test_name2"),
+#'    id = 1:2,
+#'    event_name = c("Visit 1", "Visit 1"),
+#'    item_group = c("Test_group", "Test_group2"),
+#'    form_repeat = c(1, 1),
+#'    item_name = c("Test_item", "Test_item2"),
+#'    edit_date_time = rep("2023-11-05 01:26:00", 2),
+#'    timestamp = rep("2024-02-05 01:01:01", 2)
 #'   ) |>
 #'    dplyr::as_tibble()
 #'   DBI::dbWriteTable(con, "all_review_data", review_data)
-#'   db_get_review(temp_path, subject = "Test_name", form = "Test_group")
+#'   db_get_review(temp_path, id = 1L)
+#'   db_get_review(temp_path, subject_id = "Test_name2")
 #' })
 #' 
 db_get_review <- function(
     db_path, 
-    subject = review_row$subject_id, 
-    form = review_row$item_group,
-    db_table = "all_review_data",
-    slice_vars = c("timestamp", "edit_date_time"),
-    group_vars = c("subject_id", "event_name", "item_group",
-                   "form_repeat", "item_name")
+    ...,
+    db_table = "all_review_data"
 ){
   stopifnot(file.exists(db_path))
-  stopifnot(is.character(subject))
-  stopifnot(is.character(form))
   db_temp_connect(db_path, {
-    sql <- "SELECT * FROM ?db_table WHERE subject_id = ?id AND item_group = ?group;"
-    query <- DBI::sqlInterpolate(con, sql, db_table = db_table[1], 
-                                 id = subject[1], group = form[1])
-    DBI::dbGetQuery(con, query) |> 
-      db_slice_rows(slice_vars = slice_vars, group_vars = group_vars) |> 
+    fields <- ...names()
+    sql <- paste("SELECT * FROM ?db_table WHERE", paste0(fields, " = $", fields, collapse = " AND "))
+    query <- DBI::sqlInterpolate(con, sql, db_table = db_table[1])
+    rs <- DBI::dbSendQuery(con, query)
+    DBI::dbBind(rs, params = data.frame(...))
+    df <- 
+      DBI::dbFetch(rs) |> 
       dplyr::as_tibble()
+    DBI::dbClearResult(rs)
+    df
```

Comment:
Clever construction. I did not know the `...names()` notation before. However, I am wondering if that is not too flexible in our use case. What if we only select based on 'id'? It would make this function easier to check and to maintain. 

Furthermore, currently the function requires at least one non-defined column in `...` to run (otherwise you'll get an error: `Error: near "=": syntax error`). It would be nice to throw an informative error before attempting this, and to describe that at least one argument is needed.

Reply:
So I did think about checking that `...` corresponded to the column names for the table in the DB, but I don't really want to run that check in the application itself. I suppose I could add an argument `check_names = TRUE` and we could set this as false in the app.

Reply:
For the time being I will relax the restrictions where if `...` is left empty the function will return the whole table.

Reply:
With my second comment, I meant adding something like a simple `stopifnot()` with an informative message, testing if `...names()` is not `NULL`. The SQL statement will already error if any name is incorrect, and that gives a fairly clear message I think (after a quick check).